### PR TITLE
vcpkg: init at 2023.10.19

### LINKED ISSUE #151096: Package Request: Add vcpkg
**Project description**

vcpkg is a package manager for C++ and many large and very small libraries are packaged there. Many of my projects use it for CMake configuration and for its toolchain file that automatically download and build libraries.

I want to setup dev environment with nix and this is one of the packages I'm looking for that I've been building manually (since the brew packaged vcpkg is not well configured for manifest files support)

**Metadata**

* homepage URL: https://vcpkg.io/
* source URL: https://github.com/microsoft/vcpkg
* license: MIT
* platforms: unix, linux, darwin, windows

### PR DESCRIPTION
**Pull request by gracicot**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/151096

This PR packages vcpkg, a C++ package manager: https://vcpkg.io/

The goal of this package is to make vcpkg usable inside a nix environment as developer environment, and behave as closely as possible to other package managers such as pip and npm.

To have reproducible packages that uses vcpkg to be built, we would need a tool such as `vcpkg2nix`, but projects that work with vcpkg can normally skip vcpkg and uses packages from nix directly.

I will kindly ask for some help to make this PR ready to merge, I really intent to make this an official package but I'm not sure what's left to be done, and how to test the package.

This is based on the work of @Guekka 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/fmt/default.nix
Comment (gracicot) [inline pkgs/development/libraries/fmt/default.nix:73]: This might not be required as I opened a PR to fix vcpkg upstream: https://github.com/microsoft/vcpkg-tool/pull/1168

**Comment by gracicot** [comment]
Hi @AndersonTorres ! I updated my PR. I now ensured it also works on linux, and I cleaned up the code a bit, I think it follows the contributing guidelines. I was wondering what would be required to make the PR not a draft anymore. Thank you!

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/cmakerc/default.nix:6]:
```suggestion
}:

stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/cmakerc/default.nix:13]:
```suggestion
    rev = finalAttrs.version;
```

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/cmakerc/default.nix:21]:
```suggestion
  installPhase = ''
    runHook preInstall
  
    install CMakeRC.cmake -DT $out/share/cmakerc/cmakerc-config.cmake
 
    runHook postInstall
  '';
```

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/cmakerc/default.nix:28]:
```suggestion
  meta = {
    description = "A Resource Compiler in a Single CMake Script";
    homepage = "https://github.com/vector-of-bool/cmrc";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ guekka ];
    platforms = lib.platforms.all;
  };
```

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/cmakerc/default.nix:2]: Does this software need GCC?

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:35]:
```suggestion
```

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:55]:
```suggestion
  patches = [ change-lock-location.patch ];
```

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:65]:
```suggestion
  postFixup = let
  # These are the most common binaries used by vcpkg
  # Extra binaries can be added via overlay when needed
  runtimeDeps = [
    cacert
    cmake
    git
    gzip
    meson
    ninja
    openssh
    python3
    zip
    zstd
  ];
in ''
  wrapProgram $out/bin/vcpkg --prefix PATH ${lib.makeBinPath runtimeDeps}
'';
```

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:72]:
```suggestion
  meta = {
    description = "Components of microsoft/vcpkg's binary";
    homepage = "https://github.com/microsoft/vcpkg-tool";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ guekka gracicot ];
    platforms = lib.platforms.all;
  };
```

File: pkgs/development/tools/vcpkg/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg/default.nix:9]:
```suggestion

stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/tools/vcpkg/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg/default.nix:16]:
```suggestion
    rev = finalAttrs.version;
```

File: pkgs/development/tools/vcpkg/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg/default.nix:31]: Why?

File: pkgs/development/tools/vcpkg/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg/default.nix:60]:
```suggestion
  meta = {
    description = "C++ Library Manager";
    homepage = "https://vcpkg.io/";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ guekka gracicot ];
    platforms = lib.platforms.all;
  };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Next part will be organize the commit stack.

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:18]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, cacert
, cmake
, cmakerc
, curl
, fmt
, git
, gzip
, makeWrapper
, meson
, ninja
, openssh
, python3
, zip
, zstd
}:
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/cmakerc/default.nix
Comment (gracicot) [reply pkgs/development/libraries/cmakerc/default.nix:2]: I don't think so

File: pkgs/development/tools/vcpkg/default.nix
Comment (gracicot) [reply pkgs/development/tools/vcpkg/default.nix:32]: Let me try to fix it

File: pkgs/development/libraries/cmakerc/default.nix
Comment (AndersonTorres) [reply pkgs/development/libraries/cmakerc/default.nix:2]: Then use `stdenvNoCC` instead of `stdenv`.

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:45]: Fix the indentation.

(Sorry, it is harder to write code ~~outside Emacs~~ in the GitHub web interface.)

**Review by AndersonTorres — DISMISSED** [review]
Remember to
- Create a commit for each new maintainer inserted in maintainers file;
- Create a commit for each package added
  - The commits should be topologically sorted (if X depends of Y, then X should be added after Y).

**Comment by gracicot** [comment]
I'm waiting for vcpkg-tool to be released to avoid adding the `fmt_10_0` package before continuing

File: pkgs/development/libraries/fmt/default.nix
Comment (marcusramberg) [reply pkgs/development/libraries/fmt/default.nix:73]: I see that pr got merged. :)

Comment (gracicot) [reply pkgs/development/libraries/fmt/default.nix]: I decided to apply the patch to the current version instead of waiting, since it was applied upsteam :)

File: pkgs/development/tools/vcpkg/default.nix
Comment (gracicot) [inline pkgs/development/tools/vcpkg/default.nix:34]: I'm not very proud of this code right there, but it actually works. I know there must be a better way. Can someone with a better knowledge of nix lang help me making this code less horrible? Thanks.

Comment (AndersonTorres) [reply pkgs/development/tools/vcpkg/default.nix:34]: I have a feeling you should try `substituteAll`.

**Comment by gracicot** [comment]
Should be all good now

**Comment by marcusramberg** [comment]
> Should be all good now
Still marked as work in progress btw.

**Comment by marcusramberg** [comment]
Tested nixpgs-review on aarch64-darwin and it builds fine. vcpkg list seems to want to write to the nix store tho:

`[nix-shell:~/.cache/nixpkgs-review/pr-251465]$ vcpkg list
create_directories("/nix/store/0nj7wqcs97i4kchw9h8qfh609kwfzzbl-vcpkg-2023.08.09/share/vcpkg/installed"): Permission denied
`

**Comment by gracicot** [comment]
@marcusramberg 

I left the default behaviour when you use vcpkg outside of manifest mode, so it will try to write into its root folder. You can send the argument to specifiy the install root:
```sh
$ vcpkg install fmt --x-install-root=$HOME/vcpkg_installed
```

When using vcpkg in manifest mode, you don't need to do that since it will automatically install in the build dir of your project.

I could make it so it installs into the mutable root in the user directory, but I figure we should not. The reason is that no other package manager in nix do that. Doing `npm install -g yarn` will fail with nix with an error that the nix store is not writable.

Python's pip will also fail if you try to install a package globally.

vcpkg's `install` command defaults to install globally, which should fail according to other package manager's behaviour.

**Review by marcusramberg — APPROVED** [review]
lgtm :)

File: pkgs/development/tools/vcpkg-tool/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg-tool/default.nix:18]: Use the finalAttrs pattern here too.

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:499]: Since we are now in the _by-name era_, can you move this file `development/libraries/cmakerc/default.nix` to `by-name/cm/cmakerc/package.nix`?

```suggestion
```

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:808]: Since we are now in the by-name era, can you move this file `development/tools/vcpkg-tool/default.nix` to `by-name/vc/vcpkg-tool/package.nix`?

```suggestion
  vcpkg-tool = callPackage ../by-name/vc/vcpkg-tool/package.nix { 
    fmt = fmt_10; 
  };
```

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:810]: Since we are now in the by-name era, can you move this file `development/tools/vcpkg/default.nix` to `by-name/vc/vcpkg/package.nix`?

```suggestion
```

File: pkgs/development/tools/vcpkg/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/vcpkg/default.nix:31]: don't forget runHook

**Review by AndersonTorres — DISMISSED** [review]

File: pkgs/development/tools/vcpkg/default.nix
Comment (gracicot) [reply pkgs/development/tools/vcpkg/default.nix:31]: Thanks for this one, I didn't know they had to be called explicitly. I don't see how to properly setup an install phase anywhere in the docs, the only thing I found in the manual is [this](https://nixos.org/manual/nixpkgs/stable/#ssec-install-phase), which tell very little information about how the install phase work and how to write one.

I found the relevant information here: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

**Comment by gracicot** [comment]
It should be all good now. Note that vcpkg needs CMake 3.27 and only 3.26 is packaged as far as I can see. This will cause vcpkg to download a pre-compiled cmake. It works on macos, it works on Linux when nix packages are used outside of nixos, but I doubt it will work on nixos. I see that #261427 has been merged, but testing staging seems to be taking a lot of time (I'm 9 hours in getting it built)

**Comment by AndersonTorres** [comment]
Cmake 3.27 should be on the way from staging...

https://github.com/NixOS/nixpkgs/pull/261427

**Comment by gracicot** [comment]
I was able to tests with CMake 3.27 and it works flawlessly, it doesn't attempt to download its own version anymore. It was probably overkill but I tried this PR with the whole staging branch.

**Comment by Guekka** [comment]
This looks good to me @gracicot 
However, I see that you have removed my patching system. This means we have no way to disable packages that will fail to build. Unless I'm missing something?

**Comment by gracicot** [comment]
> This looks good to me @gracicot However, I see that you have removed my patching system. This means we have no way to disable packages that will fail to build. Unless I'm missing something?

For now I would leave the failing packages as is. The patching system made it possible to ignore certain packages when attempting to install it, but it would simply result in CMake failing to find the packages in many projects since system packages are disabled by default when using vcpkg in manifest mode.

In the case where vcpkg cannot install a particular dependency, I think per project package overrides would be the way to go.

**Comment by Guekka** [comment]
> For now I would leave the failing packages as is. As far as I understood the patching system made it possible to ignore certain packages when attempting to install it, but it would simply result in CMake failing to find the packages in many projects since system packages are disabled by default when using vcpkg in manifest mode.
Unless I'm mistaken, using the Nix Qt + ignoring vcpkg Qt worked correctly on my side

> In the case where vcpkg cannot install a particular dependency, I think per project package overrides would be the way to go.
That's exactly my point, but how? What toggle do we leave to the user?

**Comment by gracicot** [comment]
> Unless I'm mistaken, using the Nix Qt + ignoring vcpkg Qt worked correctly on my side

Then Qt don't need to be installed by vcpkg for this project when compiled in a partiacular nix environment, and the system's package manager works for its needs. Does it matter if a vcpkg package fails to install in the case a user decide to use nix's package?

> That's exactly my point, but how? What toggle do we leave to the user?

vcpkg support [overlay ports](https://learn.microsoft.com/en-us/vcpkg/concepts/overlay-ports) in `vcpkg-configuration.json`, CMake variable/environment variable `VCPKG_OVERLAY_PORTS` and vcpkg command line `--overlay-ports`. Users can simply add an overlay ports that disable Qt, and set the environment variable in the nix shell to use them. A custom registry that overrides Qt would also do the trick.

Actually, even though I wasn't able to make `gettext` and `qt` to work using vcpkg, it doesn't mean someone else won't be able to make them work and document it. We should leave that to the user. Replacing certain ports also won't help in the case users configure their default registry to point directly on vcpkg's repo like some projects do.

**Comment by gracicot** [comment]
Is anything else required for this to go? Design-wise I tried to make it as close as possible to vanilla and other package managers, but if something looks off I'm willing to do changes. Let me know if anything needs to be done.

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:503]:
```suggestion
```

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:818]:
```suggestion
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (marcusramberg) [reply pkgs/top-level/all-packages.nix:818]: I thought the idea of the by-name namespace was that these packages will be included automatically in all-packages , why do we need to add them here when they don't add any attributes?

**Review by AndersonTorres — APPROVED** [review]
Ping me after 24h to merge

**Comment by gracicot** [comment]
@AndersonTorres gentle ping :)

**Comment by AndersonTorres** [comment]
I have no commit rights now. I will put this on the list:

https://discourse.nixos.org/t/prs-already-reviewed/2617/

File: pkgs/by-name/vc/vcpkg/package.nix
Comment (pbsds) [inline pkgs/by-name/vc/vcpkg/package.nix:28]:
```suggestion
        --x-downloads-root="$vcpkg_writable_path"/downloads \
        --x-buildtrees-root="$vcpkg_writable_path"/buildtrees \
        --x-packages-root="$vcpkg_writable_path"/packages \
```

File: pkgs/by-name/vc/vcpkg-tool/package.nix
Comment (pbsds) [inline pkgs/by-name/vc/vcpkg-tool/package.nix:60]:
```suggestion
    ] + extraRuntimeDeps;
```

File: pkgs/by-name/vc/vcpkg-tool/package.nix
Comment (pbsds) [inline pkgs/by-name/vc/vcpkg-tool/package.nix:16]:
```suggestion
, zstd
, extraRuntimeDeps ? []
```

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 251465` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>cmakerc</li>
    <li>vcpkg</li>
    <li>vcpkg-tool</li>
  </ul>
</details>

LGTM, just some nits

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 251465` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>cmakerc</li>
    <li>vcpkg</li>
    <li>vcpkg-tool</li>
  </ul>
</details>

**Pull request merged**
